### PR TITLE
ref(inject): Inject code snippet at the start

### PR DESCRIPTION
**Pull request by loewenheim**
This injects the code snippet as early in the file as possible, which is to say, after an initial block of `"use …";` statements, if there is one. Since those statements must itself be at the start of the file (except for empty lines and comments), we skip empty lines and comments and then any number of `"use …";` statements before injecting. For example, the file
```javascript
// a
// comment
// block

// another
// comment
// block

'use strict';
"use server";
function t(t) {
  return '[object Object]' === Object.prototype.toString.call(t);
}
//# sourceMappingURL=/path/to/sourcemap
```
would be modified to
```javascript
// a
// comment
// block

// another
// comment
// block

'use strict';
"use server";
<CODE_SNIPPET>[<debug_id>]
function t(t) {
  return '[object Object]' === Object.prototype.toString.call(t);
}
//# debugId=<debug_id>
//# sourceMappingURL=/path/to/sourcemap
```

Modifying the minified source file in this way requires us to fiddle with the sourcemap as well since every line number below the insertion point has now increased by one. Thus, we simply add `;` at the start of the sourcemap's mapping string to denote the code snippet's empty line mapping.

Question: Is it sound to just always add the empty mapping at the beginning? If any of the lines before the injection point had mappings, they would get scrambled by this change, but since they're only comments and `"use …";`, can they even have nontrivial mappings?

### CODE REVIEW

**Comment by loewenheim** [comment]
Picking this up again because it turned out to be necessary in an internal test.

File: src/utils/sourcemaps/inject.rs [hunk 1]
```diff
@@ -121,14 +165,35 @@ pub fn fixup_js_file(js_contents: &mut Vec<u8>, debug_id: DebugId) -> Result<()>
...
+    let mut js_lines = js_lines.into_iter().peekable();
+
+    // Write comments and empty lines at the start back to the file
+    while let Some(comment_or_empty) =
+        js_lines.next_if(|line| line.trim().is_empty() || line.trim().starts_with("//"))
+    {
+        writeln!(js_contents, "{comment_or_empty}")?;
     }
 
+    // Write one "use strict" statement back to the file, if there is one
```
Comment (kamilogorek) [inline src/utils/sourcemaps/inject.rs:177]: Did you consult the other pragmas with SDK team? Maybe we can just just append all lines starting with `'` or `"`, just like we do in the check above? (They are all ignored by the parsers anyway)

**Comment by kamilogorek** [comment]
Looks good to me overall.

**Comment by loewenheim** [comment]
One thing I definitely still need to handle is embedded sourcemaps.

File: src/utils/sourcemaps/inject.rs [hunk 2]
```diff
@@ -93,23 +99,69 @@ impl fmt::Display for InjectReport {
     }
 }
 
-/// Appends the following text to a file:
+/// Fixes up a minified JS source file with a debug id.
+///
+/// This changes the source file in several ways:
+/// 1. The source code snippet
```
Comment (kamilogorek) [inline src/utils/sourcemaps/inject.rs:105]: TODO: Add shebang support `#!/usr/bin/env node`

File: src/utils/sourcemaps/inject.rs [hunk 3]
```diff
@@ -121,21 +173,70 @@ pub fn fixup_js_file(js_contents: &mut Vec<u8>, debug_id: DebugId) -> Result<()>
...
     Ok(())
 }
 
+/// Replaces a JS file's source mapping url with a new one.
+///
+/// Only the bottommost source mapping url comment will be updated. If there
+/// are no source mapping url comments in the file, this is a no-op.
+pub fn replace_sourcemap_url(js_contents: &mut Vec<u8>, new_url: &str) -> Result<()> {
+    let js_lines: Result<Vec<String>, _> = js_contents.lines().collect();
+    let js_lines = js_lines?;
```
Comment (kamilogorek) [inline src/utils/sourcemaps/inject.rs:216]: Maybe `?` after collect will work. To be checked.

File: src/utils/sourcemaps.rs [hunk 4]
```diff
@@ -814,6 +839,11 @@ impl SourceMapProcessor {
                     }
                 };
 
+                inject::insert_empty_mapping(&mut sourcemap_file.contents).context(format!(
```
Comment (kamilogorek) [inline src/utils/sourcemaps.rs:842]: Create `Sourcemap` struct and pass it through `insert_empty_line`, then `fixup` and then finally serialize and store in the original file.

**Comment by kamilogorek** [comment]
TODO: Handle when no sourcemap file is available locally. In this case, we should append at the end of the file instead, as sourcemap cannot have its mappings fixed up.

File: src/utils/sourcemaps.rs (hunk 4 above)
Comment (loewenheim) [reply src/utils/sourcemaps.rs]: `fixup_sourcemap` needs access to the bytes to hash them into a debug id. I think I'll just leave it as it is.

**Comment by loewenheim** [comment]
This now also handles the edge case of a sourcemap not being available, by essentially falling back to the status quo behavior. I also refactored `inject_debug_ids` some; the double loop is no longer necessary.

**Review by kamilogorek — APPROVED** [review]

**Review by kamilogorek — APPROVED** [review]

**Pull request merged**